firebase-ai-friendly-meals/android/app/src/main/java/com/google/firebase/example/friendlymeals/data/datasource/DatabaseRemoteDataSource.kt

```diff
@@ -0,0 +1,238 @@
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.Query.Direction.DESCENDING
+import com.google.firebase.firestore.SetOptions
+import com.google.firebase.firestore.pipeline.Expression.Companion.field
+import com.google.firebase.firestore.toObject
+import kotlinx.coroutines.tasks.await
+import javax.inject.Inject
+
+class DatabaseRemoteDataSource @Inject constructor(
+    private val firestore: FirebaseFirestore
+) {
+    fun addUser(user: User) {
+        firestore.collection(USER_COLLECTION).add(user)
+    }
+
+    suspend fun addRecipe(recipe: Recipe): String {
+        val recipeRef = firestore.collection(RECIPE_COLLECTION).document()
+        val finalRecipe = recipe.copy(id = recipeRef.id)
+        recipeRef.set(finalRecipe).await()
+        return recipeRef.id
+    }
+
+    suspend fun getRecipe(recipeId: String): Recipe? {
+        return firestore
+            .collection(RECIPE_COLLECTION)
+            .document(recipeId)
+            .get().await().toObject()
+    }
+
+    suspend fun getAllRecipes(): List<RecipeListItem> {
+        return firestore.collection(RECIPE_COLLECTION)
+            .get()
+            .await()
+            .toObjects(RecipeListItem::class.java)
+    }
+
+    suspend fun addTags(tagNames: List<String>) {
+        val normalizedTags = tagNames
+            .map { it.trim() }
+            .distinct()
+
+        val batch = firestore.batch()
+        val tagsCollection = firestore.collection(TAG_COLLECTION)
+
+        normalizedTags.forEach { tagName ->
+            val tagRef = tagsCollection.document(tagName)
+
+            val data = hashMapOf(
+                NAME_FIELD to tagName,
+                TOTAL_RECIPES_FIELD to FieldValue.increment(1)
+            )
+
+            batch.set(tagRef, data, SetOptions.merge())
+        }
+
+        batch.commit().await()
+    }
+
+    suspend fun getPopularTags(): List<Tag> {
+        return firestore.collection(TAG_COLLECTION)
+            .orderBy(TOTAL_RECIPES_FIELD, DESCENDING)
+            .limit(10)
+            .get()
+            .await()
+            .toObjects(Tag::class.java)
+    }
+
+    suspend fun setReview(review: Review) {
+        val recipeRef = firestore
+            .collection(RECIPE_COLLECTION)
+            .document(review.recipeId)
+
+        val reviewRef = recipeRef
+            .collection(REVIEW_SUBCOLLECTION)
+            .document("${review.recipeId}_${review.userId}")
+
+        reviewRef.set(review).await()
+
+        val newAvg = getAverageRatingForRecipe(review.recipeId)
+        recipeRef.update(AVERAGE_RATING_FIELD, newAvg).await()
+    }
+
+    private suspend fun getAverageRatingForRecipe(recipeId: String): Double {
+        val collection = firestore.collection(RECIPE_COLLECTION)
+            .document(recipeId)
+            .collection(REVIEW_SUBCOLLECTION)
+
+        val snapshot = collection.aggregate(AggregateField.average(RATING_FIELD))
```

Comment:
nit: This is using the old aggregate API. We should use the new pipelines one, which will look like [this](https://github.com/firebase/snippets-android/pull/661/changes#diff-b8c9909359f314d46c7e39e4ccfccfbbd8f553f9749245cc345a263be53d0ce2R1432).

```kotlin
val results = db.pipeline()
    .collection(RECIPE_COLLECTION)
    .document(recipeId)
    .collection(REVIEW_SUBCOLLECTION)
    .aggregate(
        AggregateStage.withAccumulators(
            AggregateFunction.average(RATING_FIELD).alias("avg_rating")
        )
    ).execute()
```